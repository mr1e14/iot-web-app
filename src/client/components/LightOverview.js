import React from "react";
import axios from "axios";
import Grid from "@material-ui/core/Grid";
import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";
import LightBrightnessSlider from "./LightBrigtnessSlider";
import IconButton from "@material-ui/core/IconButton";
import MaterialIcon from "./MaterialIcon";
import { Link } from "react-router-dom";

const defaultBackgroundColor = "#1a1a1a";
const labelColor = "#ffffff";

class LightOverview extends React.Component {
  constructor(props) {
    super(props);
    this.state = { brightness: null };
  }

  handleToggleClick = event => {
    this.setState({ on: !this.state.on }, () =>
      axios.post("/api/lights/updateLightData", {
        id: this.props.id,
        ...this.state
      })
    );
  };

  handleBrightnessChange = value => {
    this.setState({ brightness: value }, () =>
      axios.post("/api/lights/updateLightData", {
        id: this.props.id,
        ...this.state
      })
    );
  };

  componentDidMount() {
    fetch(`/api/lights/getLightDataById/${this.props.id}`)
      .then(res => res.json())
      .then(res => this.setState({ ...res.lightData }));
  }

  render() {
    const { on, connected, brightness, color, name, _id } = this.state;
    const { classes, id } = this.props;
    return (
      <Grid
        style={{
          background: `linear-gradient(45deg, ${color ||
            defaultBackgroundColor} 35%, rgb(242,242,242) 90%)`,
          opacity: on ? 1 : 0.2,
          cursor: connected ? "pointer" : "no-drop"
        }}
        container
        alignItems="center"
        className={classes.lightRow}
      >
        <Grid item xs={10}>
          <Button
            component={Link}
            to={`/light/${id}`}
            variant="contained"
            color="primary"
            fullWidth
            disabled={!connected}
            style={{
              color: labelColor
            }}
          >
            <Typography>{name}</Typography>
          </Button>
        </Grid>
        <Grid item xs={2}>
          <IconButton
            color="primary"
            disabled={!connected}
            onClick={this.handleToggleClick}
          >
            <MaterialIcon
              iconName={
                this.state.connected ? "power_settings_new" : "power_off"
              }
            />
          </IconButton>
        </Grid>
        <Grid item xs={12}>
          {brightness !== null ? (
            <LightBrightnessSlider
              on={on}
              connected={connected}
              brightness={brightness}
              handleChange={this.handleBrightnessChange}
            />
          ) : null}
        </Grid>
      </Grid>
    );
  }
}

export default LightOverview;
