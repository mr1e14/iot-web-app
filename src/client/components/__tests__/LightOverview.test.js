import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

jest.mock("axios");

import LightOverview from "../LightOverview";
import Grid from "@material-ui/core/Grid";
import Button from "@material-ui/core/Button";
import IconButton from "@material-ui/core/IconButton";

jest.mock("react-router-dom", () => ({
  Link: "button"
}));

const exampleClasses = {
  lightRow: "lightRow"
};

const dataEnabled = {
  id: "1",
  name: "Bedroom",
  connected: true,
  on: true,
  brightness: 75,
  effect: null,
  color: "#42f48f"
};

const dataDisabled = {
  id: "2",
  name: "Bedroom",
  connected: true,
  on: false,
  brightness: 25,
  effect: null,
  color: "#42f48f"
};

const dataDisconnected = {
  id: "3",
  name: "Bedroom",
  connected: false,
  on: undefined,
  brightness: undefined,
  effect: undefined,
  color: undefined
};

describe("LightOverview", () => {
  let tree;
  let component;
  let wrapper;
  const loadData = wrapper => {
    global.fetch().then(data => data.json()).then(data => wrapper.setState({...data}))
  }
  describe("when light is enabled", () => {
    beforeEach(() => {
      global.fetch = jest.fn(() => Promise.resolve({json: () => Promise.resolve(dataEnabled)}));
      component = renderer.create(
        <LightOverview
          classes={exampleClasses}
          id="1"
        />
      );
      tree = component.toJSON();
      wrapper = mount(
        <LightOverview
          classes={exampleClasses}
          id="1"
        />
      );
      loadData(wrapper);
      wrapper.update();
    });
    it("renders correctly", () => {
      expect(tree).toMatchSnapshot();
    });
    it("has all elements enabled", () => {
      expect(wrapper.find(IconButton).prop("disabled")).toBe(false);
      expect(wrapper.find(Button).prop("disabled")).toBe(false);
      expect(wrapper.find("Slider").prop("disabled")).toBe(false);
    });
    it("has a Grid container with correct attributes", () => {
      const gridStyle = wrapper
        .find(Grid)
        .at(0)
        .prop("style");
      expect(gridStyle).toHaveProperty(
        "background",
        "linear-gradient(45deg, #42f48f 35%, rgb(242,242,242) 90%)"
      );
      expect(gridStyle).toHaveProperty("opacity", 1);
      expect(gridStyle).toHaveProperty("cursor", "pointer");
    });
    it("has a main button with correct text", () => {
      expect(wrapper.find(Button).text()).toEqual("Bedroom");
    });
    it("has a working toggle button", () => {
      wrapper.find(IconButton).simulate("click");
      expect(wrapper.find("LightOverview").instance().state.on).toBe(false);
    });
    it("has an iconButton with correct icon", () => {
      expect(wrapper.find("MaterialIcon").prop("iconName")).toEqual(
        "power_settings_new"
      );
    });
    it("has a slider which can adjust brightness", () => {
      wrapper.find("Slider").props().onChange({}, 50);
      expect(wrapper.instance().state.brightness).toBe(50);
    })
  });
  describe("when light is disabled", () => {
    beforeEach(() => {
      global.fetch = jest.fn(() => Promise.resolve({json: () => Promise.resolve(dataDisabled)}));
      component = renderer.create(
        <LightOverview
          classes={exampleClasses}
          id="2"
        />
      );
      tree = component.toJSON();
      wrapper = mount(
        <LightOverview
          classes={exampleClasses}
          id="2"
        />
      );
      loadData(wrapper);
      wrapper.update();
    });
    it("renders correctly", () => {
      expect(tree).toMatchSnapshot();
    });
    it("has both buttons enabled", () => {
      expect(wrapper.find(IconButton).prop("disabled")).toBe(false);
      expect(wrapper.find(Button).prop("disabled")).toBe(false);
    });
    it("has a disabled Slider", () => {
      expect(wrapper.find("Slider").prop("disabled")).toBe(true);
    });
    it("has a Grid container with correct attributes", () => {
      const gridStyle = wrapper
        .find(Grid)
        .at(0)
        .prop("style");
      expect(gridStyle).toHaveProperty(
        "background",
        "linear-gradient(45deg, #42f48f 35%, rgb(242,242,242) 90%)"
      );
      expect(gridStyle).toHaveProperty("opacity", 0.2);
      expect(gridStyle).toHaveProperty("cursor", "pointer");
    });
    it("has a main button with correct text", () => {
      expect(wrapper.find(Button).text()).toEqual("Bedroom");
    });
    it("has an iconButton with correct icon", () => {
      expect(wrapper.find("MaterialIcon").prop("iconName")).toEqual(
        "power_settings_new"
      );
    });
  });
  describe("when light is disconnected", () => {
    beforeEach(() => {
      global.fetch = jest.fn(() => Promise.resolve({json: () => Promise.resolve(dataDisconnected)}));
      component = renderer.create(
        <LightOverview
          classes={exampleClasses}
          id="3"
        />
      );
      tree = component.toJSON();
      wrapper = mount(
        <LightOverview
          classes={exampleClasses}
          id="3"
        />
      );
      loadData(wrapper);
      wrapper.update();
    });
    it("renders correctly", () => {
      expect(tree).toMatchSnapshot();
    });
    it("has all elements disabled", () => {
      expect(wrapper.find(IconButton).prop("disabled")).toBe(true);
      expect(wrapper.find(Button).prop("disabled")).toBe(true);
      expect(wrapper.find("Slider").prop("disabled")).toBe(true);
    });
    it("has a Grid container with correct attributes", () => {
      const gridStyle = wrapper
        .find(Grid)
        .at(0)
        .prop("style");
      expect(gridStyle).toHaveProperty(
        "background",
        "linear-gradient(45deg, #1a1a1a 35%, rgb(242,242,242) 90%)"
      );
      expect(gridStyle).toHaveProperty("opacity", 0.2);
      expect(gridStyle).toHaveProperty("cursor", "no-drop");
    });
    it("has a main button with correct text", () => {
      expect(wrapper.find(Button).text()).toEqual("Bedroom");
    });
    it("has an iconButton with correct icon", () => {
      expect(wrapper.find("MaterialIcon").prop("iconName")).toEqual(
        "power_off"
      );
    });
  });
});
