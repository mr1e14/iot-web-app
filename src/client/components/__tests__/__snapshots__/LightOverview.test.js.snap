// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LightOverview when light is disabled renders correctly 1`] = `
<div
  className="MuiGrid-root lightRow MuiGrid-container MuiGrid-align-items-xs-center"
  style={
    Object {
      "background": "linear-gradient(45deg, #1a1a1a 35%, rgb(242,242,242) 90%)",
      "cursor": "no-drop",
      "opacity": 0.2,
    }
  }
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled MuiButton-fullWidth Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "color": "#ffffff",
        }
      }
      tabIndex={-1}
      to="/light/2"
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        />
      </span>
    </button>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
  >
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <span
          aria-hidden={true}
          className="material-icons MuiIcon-root md-28"
        >
          power_off
        </span>
      </span>
    </button>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  />
</div>
`;

exports[`LightOverview when light is disconnected renders correctly 1`] = `
<div
  className="MuiGrid-root lightRow MuiGrid-container MuiGrid-align-items-xs-center"
  style={
    Object {
      "background": "linear-gradient(45deg, #1a1a1a 35%, rgb(242,242,242) 90%)",
      "cursor": "no-drop",
      "opacity": 0.2,
    }
  }
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled MuiButton-fullWidth Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "color": "#ffffff",
        }
      }
      tabIndex={-1}
      to="/light/3"
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        />
      </span>
    </button>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
  >
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <span
          aria-hidden={true}
          className="material-icons MuiIcon-root md-28"
        >
          power_off
        </span>
      </span>
    </button>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  />
</div>
`;

exports[`LightOverview when light is enabled renders correctly 1`] = `
<div
  className="MuiGrid-root lightRow MuiGrid-container MuiGrid-align-items-xs-center"
  style={
    Object {
      "background": "linear-gradient(45deg, #1a1a1a 35%, rgb(242,242,242) 90%)",
      "cursor": "no-drop",
      "opacity": 0.2,
    }
  }
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled MuiButton-fullWidth Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "color": "#ffffff",
        }
      }
      tabIndex={-1}
      to="/light/1"
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        />
      </span>
    </button>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
  >
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <span
          aria-hidden={true}
          className="material-icons MuiIcon-root md-28"
        >
          power_off
        </span>
      </span>
    </button>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  />
</div>
`;
